source/opt/inline_pass.cpp

```diff
@@ -0,0 +1,430 @@
+      ir::Operand(spv_operand_type_t::SPV_OPERAND_TYPE_ID,
+                  std::initializer_list<uint32_t>{ptrId}));
+  std::unique_ptr<ir::Instruction> newLoad(
+      new ir::Instruction(SpvOpLoad, typeId, resultId, load_in_operands));
+  (*bp)->AddInstruction(std::move(newLoad));
+}
+
+// Generate callee code into newBlocks to be inlined for the function call at
+// call_ii. Also add new function variables into caller func
+
+void InlinePass::GenInlineCode(
+    std::vector<std::unique_ptr<ir::BasicBlock>>* newBlocks,
+    std::vector<std::unique_ptr<ir::Instruction>>* newVars,
+    ir::UptrVectorIterator<ir::Instruction> call_ii,
+    ir::UptrVectorIterator<ir::BasicBlock> call_bi) {
+  // Map from callee id to caller id
+  std::unordered_map<uint32_t, uint32_t> callee2caller;
+  // Pre-call OpSampledImage Insts
+  std::unordered_map<uint32_t, ir::Instruction*> preCallSI;
+  // Post-call OpSampledImage Ids
+  std::unordered_map<uint32_t, uint32_t> postCallSI;
+
+  const uint32_t calleeId =
+      call_ii->GetOperand(kSpvFunctionCallFunctionId).words[0];
+  ir::Function* calleeFn = id2function_[calleeId];
+
+  // Map parameters to actual arguments
+  int i = 0;
+  calleeFn->ForEachParam(
+      [&call_ii, &i, &callee2caller](const ir::Instruction* cpi) {
+        const uint32_t pid =
+            cpi->GetOperand(kSpvFunctionParameterResultId).words[0];
+        callee2caller[pid] =
+            call_ii->GetOperand(kSpvFuncitonCallArgumentId + i).words[0];
+        i++;
+      });
+
+  // Define caller local variables for all callee variables and create map to
+  // them
+  auto cbi = calleeFn->begin();
+  auto cvi = cbi->begin();
+  while (cvi->opcode() == SpvOp::SpvOpVariable) {
+    std::unique_ptr<ir::Instruction> var_inst(new ir::Instruction(*cvi));
+    uint32_t newId = TakeNextId();
+    var_inst->SetResultId(newId);
+    callee2caller[cvi->result_id()] = newId;
+    newVars->push_back(std::move(var_inst));
+    cvi++;
+  }
+
+  // Create return var if needed
+  uint32_t returnVarId = 0;
+  const uint32_t calleeTypeId = calleeFn->GetTypeId();
+  const ir::Instruction* calleeType =
+      def_use_mgr_->id_to_defs().find(calleeTypeId)->second;
+  if (calleeType->opcode() != SpvOpTypeVoid) {
+    // find or create ptr to callee return type
+    uint32_t returnVarTypeId = FindPointerToType(calleeTypeId);
+    if (returnVarTypeId == 0) returnVarTypeId = AddPointerToType(calleeTypeId);
+    // Add return var to new function scope variables
+    returnVarId = TakeNextId();
+    std::vector<ir::Operand> in_operands;
+    in_operands.emplace_back(
+        spv_operand_type_t::SPV_OPERAND_TYPE_STORAGE_CLASS,
+        std::initializer_list<uint32_t>{uint32_t(SpvStorageClassFunction)});
+    std::unique_ptr<ir::Instruction> var_inst(new ir::Instruction(
+        SpvOpVariable, returnVarTypeId, returnVarId, in_operands));
+    newVars->push_back(std::move(var_inst));
+  }
+
+  // Clone and map callee code
+  bool prevInstWasReturn = false;
```

Comment:
Hm, I cannot think of the case where this is not up-to-date. It is initialized to false. It is only set to true when an OpReturn is seen. There are only two instructions that can follow: an OpLabel and an OpFunctionEnd. After OpLabel, it is set back to false (still up-to-date). I don't set it after OpFunctionEnd because then the iteration is complete and the variable is irrelevant. Right? Should I set it here anyway? Is there a case I am missing?